packages/medusa/src/services/gift-card.ts

```diff
@@ -0,0 +1,280 @@
+    manager,
+    giftCardRepository,
+    giftCardTransactionRepository,
+    regionService,
+    eventBusService,
+  }: InjectedDependencies) {
+    // eslint-disable-next-line prefer-rest-params
+    super(arguments[0])
+
+    this.manager_ = manager
+
+    this.giftCardRepository_ = giftCardRepository
+    this.giftCardTransactionRepo_ = giftCardTransactionRepository
+    this.regionService_ = regionService
+    this.eventBus_ = eventBusService
+  }
+
+  /**
+   * Generates a 16 character gift card code
+   * @return the generated gift card code
+   */
+  static generateCode(): string {
+    const code = [
+      randomize("A0", 4),
+      randomize("A0", 4),
+      randomize("A0", 4),
+      randomize("A0", 4),
+    ].join("-")
+
+    return code
+  }
+
+  /**
+   * @param selector - the query object for find
+   * @param config - the configuration used to find the objects. contains relations, skip, and take.
+   * @return the result of the find operation
+   */
+  async list(
+    selector: QuerySelector<GiftCard> = {},
+    config: FindConfig<GiftCard> = { relations: [], skip: 0, take: 10 }
+  ): Promise<GiftCard[]> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      let q
+      if ("q" in selector) {
+        q = selector.q
+        delete selector.q
+      }
+
+      const query: ExtendedFindConfig<
+        GiftCard,
+        QuerySelector<GiftCard>
+      > = buildQuery<QuerySelector<GiftCard>, GiftCard>(selector, config)
+
+      const rels = query.relations
+      delete query.relations
+
+      return await giftCardRepo.listGiftCards(query, rels, q)
+    })
+  }
+
+  async createTransaction(
+    data: CreateGiftCardTransactionInput
+  ): Promise<string> {
+    return await this.atomicPhase_(async (manager) => {
+      const gctRepo = manager.getCustomRepository(this.giftCardTransactionRepo_)
+      const created = gctRepo.create(data)
+      const saved = await gctRepo.save(created)
+      return saved.id
+    })
+  }
+
+  /**
+   * Creates a gift card with provided data given that the data is validated.
+   * @param giftCard - the gift card data to create
+   * @return the result of the create operation
+   */
+  async create(giftCard: CreateGiftCardInput): Promise<GiftCard> {
+    return await this.atomicPhase_(async (manager) => {
+      const giftCardRepo = manager.getCustomRepository(this.giftCardRepository_)
+
+      if (!giftCard.region_id) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Gift card is missing region_id`
+        )
+      }
+
+      // Will throw if region does not exist
+      const region = await this.regionService_.retrieve(giftCard.region_id)
+
+      const code = GiftCardService.generateCode()
+
+      const toCreate = {
+        code,
+        region_id: region.id,
+        ...giftCard,
+      }
+
+      const created = await giftCardRepo.create(toCreate)
```

Comment:
todo: Unnecessary `await`
```suggestion
const created = giftCardRepo.create(toCreate)
```